cloudwatch_metrics_collector/src/metrics_collector.cpp

```diff
@@ -101,26 +135,101 @@ int MetricsCollector::RecordMetrics(
                                                               dimensions,
                                                               this->storage_resolution_.load()};
     bool batched = metric_service_->batchData(metric_object);
 
     if (!batched) {
-      AWS_LOGSTREAM_ERROR(__func__, "Failed to record metric");
+      AWS_LOGSTREAM_ERROR(__func__, "Failed to record ROS monitoring message");
+    } else {
+      ++batched_count;
+    }
+  }
+  return batched_count;
+}
+
+int MetricsCollector::RecordMetrics(
+  metrics_statistics_msgs::msg::MetricsMessage::UniquePtr msg)
+{
+  int batched_count = 0;
+
+  AWS_LOGSTREAM_DEBUG(__func__, "Received metric with " << msg->statistics.size() << " types of statistics");
+  AWS_LOGSTREAM_DEBUG(__func__, "Recording metric with name=[" << msg->measurement_source_name << "]");
+
+  std::map<StatisticValuesType, double> statistic_values;
+  double stddev;
+  for (const auto & statistic : msg->statistics) {
+    if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_AVERAGE) {
+      statistic_values[StatisticValuesType::SUM] = statistic.data;
+    } else if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_MAXIMUM) {
+      statistic_values[StatisticValuesType::MAXIMUM] = statistic.data;
+    } else if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_MINIMUM) {
+      statistic_values[StatisticValuesType::MINIMUM] = statistic.data;
+    } else if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_STDDEV) {
+      stddev = statistic.data;
+    } else if (statistic.data_type == StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT) {
+      statistic_values[StatisticValuesType::SAMPLE_COUNT] = statistic.data;
     }
+  }
+  try {
+    double & sum = statistic_values.at(StatisticValuesType::SUM);
+    sum *= statistic_values.at(StatisticValuesType::SAMPLE_COUNT);
+  } catch (const std::out_of_range & exception) {
+    statistic_values.erase(StatisticValuesType::SUM);
+  }
+
+  std::map<std::string, std::string> dimensions;
+  dimensions["metric_type"] = msg->metrics_source;
+  dimensions["metric_unit"] = msg->unit;
+
+  // create a MetricObject with message parameters to batch
+  MetricObject metric_object(
+    msg->measurement_source_name,
+    "msec",
+    GetMetricDataEpochMillis(msg->window_start),
+    GetMetricDataEpochMillis(msg->window_stop) - GetMetricDataEpochMillis(msg->window_start),
+    statistic_values,
+    dimensions,
+    storage_resolution_.load()
+  );
 
-    batched_count++;
+  bool batched = metric_service_->batchData(metric_object);
+  if (!batched) {
+    AWS_LOGSTREAM_ERROR(__func__, "Failed to record metrics statistics message");
+  } else {
+    ++batched_count;
   }
+
+  dimensions.clear();
+  dimensions["metric_type"] = msg->metrics_source + " stddev";
+
+  MetricObject stddev_object(
+    msg->measurement_source_name,
```

Comment:
I disagree with this conversion as well. 
1. We now have 2 metrics of the same name but different units and values 
2. and unless someone chooses to pick the dimension correctly it will cause the query for these metrics to be incorrect. I believe there is no way to query for metrics that don't have dimension set. 

Instead of setting the dimension, change the metrics name to be `msg->measurement_source_name + "_stddev"` or something like that. 

Also whats the difference between `msg->measurement_source_name` vs `msg->metrics_source` ?

Reply:
`measurement_source_name` is a bigger category than `metrics_source`. `measurement_source_name` could be the name of the system or the name of a node or process ID, while `metrics_source` would then be its memory usage, CPU usage, etc.